lib/spack/spack/multimethod.py

```diff
@@ -120,32 +120,45 @@ def _get_method_by_spec(self, spec):
                 return method
         return self.default or None
 
-    def __call__(self, package_self, *args, **kwargs):
+    def __call__(self, package_or_builder_self, *args, **kwargs):
         """Find the first method with a spec that matches the
         package's spec.  If none is found, call the default
         or if there is none, then raise a NoSuchMethodError.
         """
-        spec_method = self._get_method_by_spec(package_self.spec)
+        spec_method = self._get_method_by_spec(package_or_builder_self.spec)
         if spec_method:
-            return spec_method(package_self, *args, **kwargs)
+            return spec_method(package_or_builder_self, *args, **kwargs)
         # Unwrap the MRO of `package_self by hand. Note that we can't
         # use `super()` here, because using `super()` recursively
         # requires us to know the class of `package_self`, as well as
         # its superclasses for successive calls. We don't have that
         # information within `SpecMultiMethod`, because it is not
         # associated with the package class.
-        for cls in inspect.getmro(package_self.__class__)[1:]:
+        for cls in inspect.getmro(package_or_builder_self.__class__)[1:]:
             superself = cls.__dict__.get(self.__name__, None)
+
             if isinstance(superself, SpecMultiMethod):
                 # Check parent multimethod for method for spec.
-                superself_method = superself._get_method_by_spec(package_self.spec)
+                superself_method = superself._get_method_by_spec(package_or_builder_self.spec)
                 if superself_method:
-                    return superself_method(package_self, *args, **kwargs)
+                    return superself_method(package_or_builder_self, *args, **kwargs)
             elif superself:
-                return superself(package_self, *args, **kwargs)
+                return superself(package_or_builder_self, *args, **kwargs)
+
+        # Account for the presence of `__getattr__` in the class
+        for cls in inspect.getmro(package_or_builder_self.__class__)[1:]:
+            if "__getattr__" in cls.__dict__:
+                try:
+                    superself_fn = cls.__getattr__(package_or_builder_self, self.__name__)
```

Comment:
> From looking at inspect.getmro it returns a sequence of class objects, which should all define getattr (vs. super() returning a super object which does not support getattr).

Not all objects support `__getattr__` (which is not `__getattribute__`). A simple example:
```python
Spack version 0.19.0.dev0
Python 3.8.10, Linux x86_64
>>> class A:
...     pass
... 
>>> A.__getattr__
Traceback (most recent call last):
  File "<console>", line 1, in <module>
AttributeError: type object 'A' has no attribute '__getattr__'
>>> A.__getattribute__
<slot wrapper '__getattribute__' of 'object' objects>
```
A few notes:
1. `@when` is used for functions called in many contexts, not only during installations.
2. This change "mimics" the protocol used by Python to access attributes via `self`: first try explicitly defined attributes, then fallback to `__getattr__` if it exist.

Reply:
While I can't `A().__getattr__(...)` I *can* `getattr(A(), ...)`. Is there a reason you specifically want the `__getattr__` method?

Reply:
For two reasons:
1. Unify the implementation for returning the forwarding attributes
2. Catch automatically methods not listed in `legacy_*` and accessed through `self`

Another note (maybe not needed, but to be explicit, since the Python jargon here doesn't help clarifying concepts):
1. Using `getattr(obj, 'name', default)` goes through a "protocol" that falls-back on calling `__getattr__` if no explicit method is found and `__getattr__` is defined on `self`
2. One can use `getattr(super(type(obj), obj), ...)` but the wrapper returned by `super` doesn't fall-back on `__getattr__` defined in the `type(obj` hierarchy

Reply:
If I understand:

* we already checked for definition of the attribute with `superself = cls.__dict__.get(self.__name__, None)` above and so using `getattr` would be redundant with that
* Other than that, using `getattr(package_or_builder_self, self.__name__)` would be fine: like this code, it would use the "nearest" `__getattr__` definition

I think it would be more clear to:

* Relocate the `__getattr__` behavior which fetches from the builder to an explicitly-named function (like `get_attribute_from_builder`
* At this stage, we look for *that* method rather than checking for `__getattr__` (subclass overrides of `__getattr__` should never implement behavior which searches for definitions in the builder)

Also, if `package_or_builder_self` is a `Package`, then I assume that it's not an issue if the Builder uses multimethods because:

* the default builders do not use `@when` on their methods
* therefore if a builder has this, then it was explicitly defined in the package.py file (and none of the forwarding logic is used in that case)

Reply:
I think we are talking about two different things. So, I'll try to clarify:
1. The extension in `multimethod to `__getattr__` should not be needed at this point (it was needed in an earlier implementation). I kept it since it mimics what Python does with self, so it seemed a good extension.
2. Having `__getattr__` defined in the adapter is good for point 2 in https://github.com/spack/spack/pull/30738#discussion_r985522624

Reply:
> I don't think I follow, but if you mean that this for loop can be removed with no change in behavior, then I'd prefer that. I assumed this was for retrieving a possible multimethod from the associated package when we are looking inside the builder (and vice versa).

Removed. We'll see if there are any consequences. To be clear, this should not be necessary anymore since we have this code:

https://github.com/spack/spack/blob/b3dcd5b0af745650dcb6e49af77f28f0b5700cb8/lib/spack/spack/builder.py#L147-L168

that adds the methods explicitly to the object (to deal with `super()` calls).